### PR TITLE
Ensure documents always scroll to exact browse mode position

### LINKED ISSUE #9611: Focus movement changing between focus and browse mode with UIA in Word 365
### Steps to reproduce:
1. Open Word 365 with a document longer than visually fits on screen.  Have the option to use UIA in Word checked in NVDA's advanced options.
2. At the top of the document, press NVDA+spacebar to put NVDA in Browse mode.
3. Press CONTROL+END to move to the end of the document.
4. Press NVDA+spacebar to put NVDA in Focus mode.
5. Press the up arrow.

### Actual behavior:

A) At Step 3, the screen does NOT scroll to show the current location of the cursor.
B) At Step 5, the focus jumps back to the start of the first line.  Note that if the left arrow is pressed in Step 5, the focus moves to the END of the first line.  If the right arrow or down arrow are pressed, the focus moves to a spot just under the visible text (eg if one full page is visible, the focus moves to the start of page 2).

C) At Step 3, instead of moving to the end of the document, if the focus is moved within the visible text on screen, the focus is where it would be expected to be after step 5.

### Expected behavior:
At Step 3, the screen should scroll to show the end of the document.
At Step 5, the focus should be at the end of the document (one character from the end if left arrow is pressed, on the second last line if the up arrow is pressed).

#### NVDA version:
2019.1.1 installed

#### Windows version:
Windows 10 Version: 
18362.1 
#### Name and version of other software in use when reproducing the issue:
Office 365 Version: 
16.0.11601.20184 

#### Does the issue still occur after restarting your PC?
Yes.  Also reported by a user.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
NVDA alpha-17311,0d29265e - same behaviour.

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #9611

### Summary of the issue:
When reading / navigating with browse mode in Microsoft Word via UI Automation, the document does not always scroll so that the current browse mode position is visible on screen, nor does the focus mode caret position stay in sync with the browse mode caret, especially when moving across pages in browse mode.
Browse mode currently only knows how to scroll to objects in the document, not specific text range positions. Thus, it is only able to scroll to lists, tables and graphics, and not any arbitrary text.
Also, setting the caret in Microsoft Word to a range that is on a page that is currently off screen seems to fail and or set the caret to a random point on the current page.

### Description of how this pull request fixes the issue:
* Add a scrollIntoView method on `textInfos.TextInfo` which should croll the document containing that range such that the range is visible on screen. The base implementation has no implementation and does nothing.
* Implement ScrollIntoView for IAccessible2 textInfos, UIA textInfos, Microsoft Word object model textInfos, and virtualBuffer textInfos. The VirtualBuffer textInfo implementation comes straight from the old BrowseMode logic which is to call ScrollIntoView on the deepest object at the position in the document.
* Change BrowseModeDocument's `_set_selection` to call ScrollIntoView on the given textInfo, rather than the deepest object at the position of that textInfo. This is much more accurate and will ensure that any arbitrary textRange is scrolled to. It is specifically this change, plus the implementation of ScrollIntoView for UIA TextInfo that solves the Microsoft  Word with Browse mode via UI Automation scrolling issue.

### Testing strategy:
Try build: https://ci.appveyor.com/api/buildjobs/bsjf2a8c6t9879wf/artifacts/output%2Fnvda_snapshot_try-i9611-23673%2C38a7b5ac.exe
@Qchristensen  has confirmed this pr fixes #9611 and  that it does not cause regressions in any other browse mode implementation (Firefox, Chrome etc).

### Known issues with pull request:
None known.

### Change log entries:
Bug fixes:
* Reading / navigating with browse mode in Microsoft Word via UI automation now ensures the document is always scrolled so that current browse mode position is visible, and that the caret position in focus mode correctly reflects the browse mode position. (#9611)

### Code Review Checklist:

- [ ] Pull Request description is up to date.
- [ ] Unit tests.
- [ ] System (end to end) tests.
- [ ] Manual testing.
- [ ] User Documentation.
- [ ] Change log entry.
- [ ] Context sensitive help for GUI changes.
- [ ] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

### CODE REVIEW

**Review by lukaszgo1 — COMMENTED** [review]
There are some unrelated changes in this PR. Can it be rebased on master?

**Review by seanbudd — APPROVED** [review]

File: source/virtualBuffers/__init__.py [hunk 1]
```diff
@@ -394,6 +394,17 @@ def getMathMl(self, field):
...
 		return obj.mathMl
 
+	def scrollIntoView(self):
+		# Scroll the deepest object at this point into view.
+		obj = self.NVDAObjectAtStart
+		if not obj:
+			return
+		if obj == self.obj.rootNVDAObject:
+			# However never scroll to the  document itself
+			return
```
Comment (seanbudd) [inline source/virtualBuffers/__init__.py:404]: Should these cases be logged? `if not obj` used to be

**Review by seanbudd — CHANGES_REQUESTED** [review]

**Review by seanbudd — APPROVED** [review]

**Comment by michaelDCurran** [comment]
I added back the logging for `not obj` but, the equal to rootNVDAObject 
should not be logged as this is very common and deliberately should be 
ignored.

**Comment by leonardder** [comment]
Ah, a pity I missed this. There seems to be some overlap with #9919 which I abandoned. Some thoughts:

1. This pr does not implement scrolling for MSHTML. It does support scrolling at the text range level, though.
2. I implemented an allignToTop parameter in #9919 since most API's support specifying this. It was used in browse mode because otherwise, some implementations (I don't remember which) tent to scroll all the time, even when not necessary.
3. IAccessible2 ScrollSubstringTo sometimes fails when the object is entirely out of view.

@michaelDCurran  I assume you didn't know about #9919. May be it could be helpful to have a look at it and see what can be incorporated?

**Comment by michaelDCurran** [comment]
Ah sorry, I totally missed #9919.

It seems to be more complete. I'll take a proper look and decide what to do.

**Comment by michaelDCurran** [comment]
@leonardder Although #9919 is much more complete, I believe this pr is enough of a subset for what is required for MS Word with UIA to scroll, without breaking anything else.
True that MSHTML is not implemented, but currently nothing would call it anyway. Any browse mode implementation we currently have for MSHTML is a virtualBuffer.
I'd be happy to take a look at what can be taken from #9919 at a later date. But for now I'd only be interested in real-world regressions caused by this PR.
I think the only one that might exist is scrolling in Amazon Kindle for PC -- this is the only browse mode implementation that uses IAccessible2 directly, and it is possible scrollToSubstring may have the limitation you noted in #9919. Though that would need to be tested.

**Pull request merged**